doc/developer/design/20240117_decoupled_storage_controller.md

```diff
@@ -0,0 +1,155 @@
+    installing read holds.
+  - TablesController: facilitates writes to tables. Could also be called
+    TxnController.
+  - StorageController: manages storage-flavored computation running on a
+    cluster. Think sources and sinks. This is the same as ComputeController,
+    but for running storage-flavored computation on a cluster.
+
+## Overview
+
+- We need to take apart the naturally grown responsibilities of current
+  StorageController to make it fit into a distributed, post-platform-v2 world.
+- CollectionsController and TablesController will become roughly thin-clients
+  that use internal "widgets" to do distributed coordination:
+  - For CollectionsController, that widget is persist since handles for holding
+    a since and other handles for learning about upper advancement.
+  - For TablesController, that widget is `persist-txn`. But there is more work
+    in that area, specifically we need to replace the current process-level
+    write lock.
+
+## CollectionsController
+
+Sketch of what I think we need and why:
+
+- The compute controller only needs a CollectionsController. It has a much
+  reduced interface, compared to current StorageController. Greatly reducing
+  surface area/coupling between the two.
+- We will revive an old idea where we use persist directly to learn about the
+  upper of collections: the CollectionsController does not rely on feedback
+  from the cluster to learn about uppers. It uses persist, and drives its since
+  handles forward based on that.
+- Each compute controller is given their own "instance" of a
+  CollectionsController, that they own.
+- A CollectionsController will not initially acquire since handles for _all_
+  collections but only for those in which the compute controller expresses an
+  interest.
+- In the past, we were hesitant about this approach because we wouldn't want to
+  regularly poll persist for uppers. Now, with persist pubsub, that wouldn't be
+  a problem anymore. Persist pubsub would essentially become the fabric that
+  ships upper information around a whole environment, between different
+  processes.
+
+Advantages of this approach:
+
+- Reduced coupling/much reduced surface area between ComputeController and
+  CollectionsController.
+- Makes use-case isolation/distributed processes possible!
+
+Implications:
+
+- In the limit, it can now happen that we hold `num_collection * num_clusters`
+  since handles, where before it was only `num_collection` handles.
+- More load on persist pubsub.
+- One could argue there would be more latency in learning about a new upper,
+  but I don't think that's a valid concern: persist pubsub has so far proven to
+  be low latency in practice. And there's motivation to invest into fixing
+  issues with persist pubsub because it is used in all places where persist
+  shards are being read.
+
+## TablesController
+
+The `StorageController` only needs to know about tables when sinking them, and
```

Comment:
Sorry, this was quite terse. I meant to say: We can move table-related things out into a `TablesController` _because_ the `StorageController` doesn't really do much with/to them except:

- learn about upper updates through a channel, where updates get sent to when adapter writes to a table
- acquire since holds when sinking them (their collection) to an export

For both of these use cases, the `StorageController` can be given access to a `CollectionsController`, and acquire read holds same as everyone else (same as compute and the adapter).

Reply:
This makes quite a bit of sense! Thank you for the clarification!